pkg/alibabacloud/eni/node.go

```diff
@@ -0,0 +1,396 @@
+			e, ok := rev.Resource.(*eniTypes.ENI)
+			if !ok {
+				return nil
+			}
+
+			n.enis[e.NetworkInterfaceID] = *e
+			if len(e.PrivateIPSets) <= 1 {
+				return nil
+			}
+
+			for _, ip := range e.PrivateIPSets {
+				if ip.Primary {
+					continue
+				}
+				available[ip.PrivateIpAddress] = ipamTypes.AllocationIP{Resource: e.NetworkInterfaceID}
+			}
+			return nil
+		})
+	enis := len(n.enis)
+	n.mutex.Unlock()
+
+	// An ECS instance has at least one ENI attached, no ENI found implies instance not found.
+	if enis == 0 {
+		scopedLog.Warning("Instance not found! Please delete corresponding ciliumnode if instance has already been deleted.")
+		return nil, fmt.Errorf("unable to retrieve ENIs")
+	}
+
+	return available, nil
+}
+
+// PrepareIPAllocation returns the number of ENI IPs and interfaces that can be
+// allocated/created.
+func (n *Node) PrepareIPAllocation(scopedLog *logrus.Entry) (*ipam.AllocationAction, error) {
+	l, limitsAvailable := n.getLimits()
+	if !limitsAvailable {
+		return nil, fmt.Errorf("Unable to determine limits")
+	}
+	a := &ipam.AllocationAction{}
+
+	n.mutex.RLock()
+	defer n.mutex.RUnlock()
+
+	for key, e := range n.enis {
+		if e.Type != "Secondary" {
+			continue
+		}
+		scopedLog.WithFields(logrus.Fields{
+			fieldEniID:  e.NetworkInterfaceID,
+			"ipv4Limit": l.IPv4,
+			"allocated": len(e.PrivateIPSets),
+		}).Debug("Considering ENI for allocation")
+
+		availableOnENI := math.IntMax(l.IPv4-len(e.PrivateIPSets), 0)
+		if availableOnENI <= 0 {
+			continue
+		} else {
+			a.AvailableInterfaces++
+		}
+
+		scopedLog.WithFields(logrus.Fields{
+			fieldEniID:       e.NetworkInterfaceID,
+			"availableOnEni": availableOnENI,
+		}).Debug("ENI has IPs available")
+
+		if subnet := n.manager.GetVSwitch(e.VSwitch.VSwitchID); subnet != nil {
+			if subnet.AvailableAddresses > 0 && a.InterfaceID == "" {
+				scopedLog.WithFields(logrus.Fields{
+					"vSwitchID":          e.VSwitch.VSwitchID,
+					"availableAddresses": subnet.AvailableAddresses,
+				}).Debug("Subnet has IPs available")
+
+				a.InterfaceID = key
+				a.PoolID = ipamTypes.PoolID(subnet.ID)
+				a.AvailableForAllocation = math.IntMin(subnet.AvailableAddresses, availableOnENI)
+			}
+		}
+	}
+	a.AvailableInterfaces = l.Adapters - len(n.enis) + a.AvailableInterfaces
+	return a, nil
+}
+
+// AllocateIPs performs the ENI allocation operation
+func (n *Node) AllocateIPs(ctx context.Context, a *ipam.AllocationAction) error {
+	_, err := n.manager.api.AssignPrivateIpAddresses(ctx, a.InterfaceID, a.AvailableForAllocation)
+	return err
+}
+
+// PrepareIPRelease prepares the release of ENI IPs.
+func (n *Node) PrepareIPRelease(excessIPs int, scopedLog *logrus.Entry) *ipam.ReleaseAction {
+	r := &ipam.ReleaseAction{}
+
+	n.mutex.Lock()
+	defer n.mutex.Unlock()
+
+	// Iterate over ENIs on this node, select the ENI with the most
+	// addresses available for release
+	for key, e := range n.enis {
+		if e.Type != "Secondary" {
```

Comment:
I'm assuming `Secondary` is defined in the API and it is not necessary to check for the lower case word? i.e. "secondary"